lib/timers.js

```diff
@@ -554,67 +557,75 @@ function tryOnImmediate(immediate, queue) {
   }
 }
 
+function runCallback(timer) {
+  const argv = timer._argv;
+  const argc = argv ? argv.length : 0;
+  switch (argc) {
+    // fast-path callbacks with 0-3 arguments
+    case 0:
+      return timer._callback();
+    case 1:
+      return timer._callback(argv[0]);
+    case 2:
+      return timer._callback(argv[0], argv[1]);
+    case 3:
+      return timer._callback(argv[0], argv[1], argv[2]);
+    // more than 3 arguments run slower with .apply
+    default:
+      return timer._callback.apply(timer, argv);
+  }
+}
 
-function Immediate() { }
-
-Immediate.prototype.domain = undefined;
-Immediate.prototype._onImmediate = undefined;
-Immediate.prototype._idleNext = undefined;
-Immediate.prototype._idlePrev = undefined;
 
+function Immediate() {
+  // assigning callback here in the constructor results in pessimal performance,
```

Comment:
worse than poor, it was a 10x slowdown due to optimization / deoptimization thrashing.

What were you thinking to suggest setting the fields to noop first?
That seems to allow the callback to be set without triggering the thrashing.